irr/src/CImageWriterPNG.cpp

```diff
@@ -123,70 +125,59 @@ bool CImageWriterPNG::writeImage(io::IWriteFile *file, IImage *image, u32 param)
 	default:
 		break;
 	}
-	u8 *tmpImage = new u8[image->getDimension().Height * lineWidth];
-	if (!tmpImage) {
-		os::Printer::log("PNGWriter: Internal PNG create image failure", file->getFileName(), ELL_ERROR);
-		png_destroy_write_struct(&png_ptr, &info_ptr);
-		return false;
-	}
+	std::unique_ptr<u8[]> tmpImage{new u8[img_dim.Height * lineWidth]};
 
+	u64 num_pixels = img_dim.Height * img_dim.Width;
```

Comment:
It's `u32`s, so using a u64 here is misleading, imo.
```suggestion
	auto num_pixels = img_dim.Height * img_dim.Width;
```

Reply:
I think the idea is that `Height` and `Width` require up to 32 bits so in theory the product might require 64 bits. (In order for this to work the dimensions would have to be casted before multiplication though?)

Arguably the real problem is earlier on: I'm pretty sure the dimensions should be `u16` to begin with (even though file formats like PNG technically allow larger textures; those should be checked against) because there's no way we support > 16k² textures with any reasonable standard of performance.

Reply:
>(In order for this to work the dimensions would have to be casted before multiplication though?)

Yes, that's why I find it misleading.

Reply:
> In order for this to work the dimensions would have to be casted before multiplication though?

No, in C and C++ with 64 bit integer size, small integer operands (smaller than 64 bit) are cast to signed 64-bit int before the mulitplication; see https://stackoverflow.com/a/46073296.

Reply:
Integer types are promoted to `int`, but that's 32 bit on most 64 bit systems.
Or am I overlooking something?

Reply:
For some reason I have misremembered the size of int to be 64 bit on amd64. If int takes 32 bit and I understand it correctly, the operation is done with unsigned 32 bit precision and the operands would indeed need to be casted to u64.